File: java/leetcode/src/main/java/com/leetcode/graph/Flood_Fill_733.java
package com.leetcode.graph;

public class Flood_Fill_733 {
    int[][] dirs = new int[][]{{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        int m = image.length;
        int n = image[0].length;
        boolean[][] visited = new boolean[m][n];
        dfs(image, sr, sc, image[sr][sc], color, visited);
        return image;
    }

    void dfs(int[][] image, int i, int j, int num, int color, boolean[][] visited) {
        if(visited[i][j]) return;
        visited[i][j] = true;
        image[i][j] = color;
        for (int d = 0; d < 4; d++) {
            int[] dir = dirs[d];
            int row = dir[0] + i;
            int col = dir[1] + j;
            if (row >= 0 && row < image.length && col >= 0 && col < image[0].length && !visited[row][col] && image[row][col] == num) {
                dfs(image, row, col, num, color, visited);
            }
        }
    }
}

File: js/algrithm/huahua130/Graph/Flood_Fill_733.js
/**
 * @param {number[][]} image
 * @param {number} sr
 * @param {number} sc
 * @param {number} color
 * @return {number[][]}
 */
var floodFill = function(image, sr, sc, color) {
    const dirs = [[1, 0], [-1, 0], [0, 1], [0, -1]]
    const m = image.length
    const n = image[0].length
    const visited = []
    for (let i = 0; i < m; i++) {
        visited.push(Array(n).fill(false))
    }

    const dfs = function(image, i, j, color, visited, origin) {
        visited[i][j] = true
        image[i][j] = color
        for (let dir of dirs) {
            let nr = dir[0] + i
            let nc = dir[1] + j
            if (nr >=0 && nr < m && nc >=0 && nc < n && !visited[nr][nc] && image[nr][nc] === origin) {
                dfs(image, nr, nc, color, visited, origin)
            }
        }
    }

    dfs(image, sr, sc, color, visited, image[sr][sc])

    return image
};
const image =  [[1,1,1],[1,1,0],[1,0,1]]
const res = floodFill(image, 1, 1, 2)
console.log(res)
